src/cargo/sources/git/oxide.rs

```diff
@@ -14,26 +17,12 @@ pub fn with_retry_and_progress(
     repo_path: &std::path::Path,
     config: &Config,
     cb: &mut (dyn FnMut(
-        &git::Repository,
+        &std::path::Path,
         &AtomicBool,
         &mut git::progress::tree::Item,
-    ) -> CargoResult<()>
-              + Send),
+        &mut dyn FnMut(&git::bstr::BStr),
+    ) -> CargoResult<()>),
 ) -> CargoResult<()> {
```

Comment:
Stacked Git incentivises categorization and is great to prevent many small commits that don't carry the big picture well, but in this case the category 'make tests work' is clearly too broad.

I hope it will be alright to add specific motivations to the body of the commit and help explain what's going on, because this change is indeed in the name of fixing a test (the one that produces a message with URLs tried). Documentation about the callback itself should probably be improved as well while at it.

It's now an item on the TODO list.

Reply:
imo I would prefer to see a commit that is focused only on fixing that test or the fix to be merged into the relevant earlier commits if the context that a bug introduced in this PR was fixed in this PR isn't relevant long term.  Burying production changes in a commit focused on enabling tests could make it hard to discover when spelunking the code, even if the commit has context about it.

Reply:
That makes sense and I will try to either port changes to other commits if they fit better there, or to split out new commits that are explicit about fixing one particular test or feature. TODO list updated.